### PR TITLE
added detailed opening hours to stores

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Administrators need to set more detailed opening hours and show if the store is currently opened or closed.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: project-base/app/src/Model/Store/StoreDataFactory.php [hunk 1]
```diff
@@ -46,7 +54,10 @@ public function createFromStore(Store $store): StoreData
         $storeData->city = $store->getCity();
         $storeData->postcode = $store->getPostcode();
         $storeData->country = $store->getCountry();
-        $storeData->openingHours = $store->getOpeningHours();
+        $storeData->openingHours = array_map(
+            fn (OpeningHours $openingHours): OpeningHoursData => $this->openingHourDataFactory->createFromOpeningHour($openingHours),
+            $store->getOpeningHours(),
+        );
```
Comment (grossmannmartin) [inline project-base/app/src/Model/Store/StoreDataFactory.php:60]: I would probably move this method into the factory, but I don't mind it this way.

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: project-base/app/src/Model/Store/StoreRepository.php [hunk 2]
```diff
@@ -179,6 +180,7 @@ public function getByIdEnabledOnDomain(int $id, int $domainId): Store
     {
         $store = $this->getQueryBuilder()
             ->join(StoreDomain::class, 'sd', Join::WITH, 's.id = sd.store AND sd.isEnabled = TRUE AND sd.domainId = :domainId')
+//            ->join(OpeningHours::class, 'oh', Join::WITH, 's.id = oh.store')
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/StoreRepository.php:183]: removed

File: project-base/app/config/graphql/types/ModelType/Store/OpeningHours/OpeningHours.types.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    type: object
+    config:
+        description: "Represents store opening hours"
+        fields:
+            isOpen:
+                type: "Boolean!"
+                description: "Is store currently open?"
+            dayOfWeek:
+                type: "Int!"
+                description: "Current day of the week"
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/config/graphql/types/ModelType/Store/OpeningHours/OpeningHours.types.yaml:11]: Added

File: project-base/app/src/DataFixtures/Demo/StoreDataFixture.php [hunk 4]
```diff
@@ -183,4 +200,49 @@ public function getDependencies()
...
+            $dateInterval = new DateInterval(sprintf('PT%sH', $i < 6 ? $i - 1 : $i - 4));
+
+            $openingHour->dayOfWeek = $i;
+            $openingHour->firstOpeningTime = (clone $firstOpeningTime)->add($dateInterval)->format('H:i');
+            $openingHour->firstClosingTime = $firstClosingTime->format('H:i');
+
+            if ($i < 6) {
+                $openingHour->secondOpeningTime = $secondOpeningTime->format('H:i');
+                $openingHour->secondClosingTime = (clone $secondClosingTime)->sub($dateInterval)->format('H:i');
+            }
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/DataFixtures/Demo/StoreDataFixture.php:241]: Hmmm, okay

File: project-base/app/src/Model/Store/OpeningHours/Exception/OpeningHoursException.php [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Model\Store\OpeningHours\Exception;
+
+use Throwable;
+
+interface OpeningHoursException extends Throwable
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/Exception/OpeningHoursException.php:9]: fixed

File: project-base/app/src/Model/Store/OpeningHours/OpeningHours.php [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+ */
+class OpeningHours
+{
+    /**
+     * @var int
+     * @ORM\Column(type="integer")
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="IDENTITY")
+     */
+    private int $id;
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/OpeningHours.php:22]: I have updated entity according to our framework rules

File: project-base/app/src/Model/Store/OpeningHours/OpeningHours.php [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    public function getId(): int
+    {
+        return $this->id;
+    }
+
+    /**
+     * @return \App\Model\Store\Store
+     */
+    public function getStore(): Store
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/OpeningHours.php:80]: removed

File: project-base/app/src/Model/Store/OpeningHours/OpeningHoursData.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Model\Store\OpeningHours;
+
+class OpeningHoursData
+{
+    public int $dayOfWeek = 0;
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/OpeningHoursData.php:9]: Ok

File: project-base/app/src/DataFixtures/Demo/StoreDataFixture.php [hunk 9]
```diff
@@ -183,4 +200,49 @@ public function getDependencies()
...
+        $secondOpeningTime = $demoRow[self::ATTR_OPENING_HOURS_SECOND_OPENING_TIME];
+        /** @var \DateTime $secondClosingTime */
+        $secondClosingTime = $demoRow[self::ATTR_OPENING_HOURS_SECOND_CLOSING_TIME];
+
+        for ($i = 1; $i <= 7; $i++) {
+            $openingHour = $this->openingHourDataFactory->create();
+            $dateInterval = new DateInterval(sprintf('PT%sH', $i < 6 ? $i - 1 : $i - 4));
+
+            $openingHour->dayOfWeek = $i;
+            $openingHour->firstOpeningTime = (clone $firstOpeningTime)->add($dateInterval)->format('H:i');
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/DataFixtures/Demo/StoreDataFixture.php:235]: I have rewritten it to setting exact values

File: project-base/app/src/Model/Store/OpeningHours/OpeningHoursFacade.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            'isOpen' => $this->isOpen($now, $todayOpeningHours),
+            'dayOfWeek' => $dayOfWeek,
+            'openingHoursOfDays' => array_map(static fn (OpeningHours $openingHours): array => [
+                'dayOfWeek' => $openingHours->getDayOfWeek(),
+                'firstOpeningTime' => $openingHours->getFirstOpeningTime(),
+                'firstClosingTime' => $openingHours->getFirstClosingTime(),
+                'secondOpeningTime' => $openingHours->getSecondOpeningTime(),
+                'secondClosingTime' => $openingHours->getSecondClosingTime(),
+            ], $store->getOpeningHours()),
+        ];
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/OpeningHoursFacade.php:45]: I have moved it to resolver map

File: project-base/app/src/Model/Store/OpeningHours/OpeningHoursFacade.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * @param \DateTimeImmutable $now
+     * @param \App\Model\Store\OpeningHours\OpeningHours|null $openingHours
+     * @return bool
+     */
+    private function isOpen(DateTimeImmutable $now, ?OpeningHours $openingHours): bool
+    {
+        if ($openingHours === null) {
+            return false;
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/OpeningHours/OpeningHoursFacade.php:56]: You are right

File: project-base/app/src/Model/Store/Store.php [hunk 12]
```diff
@@ -99,10 +102,11 @@ class Store implements OrderableEntityInterface
     protected Country $country;
 
     /**
-     * @var string|null
-     * @ORM\Column(type="text", nullable=true)
+     * @var \Doctrine\Common\Collections\Collection
+     * @ORM\OneToMany(targetEntity="\App\Model\Store\OpeningHours\OpeningHours", mappedBy="store", cascade={"persist", "remove"}, orphanRemoval=true, fetch="EAGER")
+     * @ORM\OrderBy({"id" = "ASC"})
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/Store.php:107]: Good catch!

File: project-base/app/src/Model/Store/Store.php [hunk 13]
```diff
@@ -292,11 +297,27 @@ public function getCountry(): Country
...
-        return $this->openingHours;
+        return $this->openingHours->getValues();
+    }
+
+    /**
+     * @return \App\Model\Store\OpeningHours\OpeningHours
+     */
+    public function getTodayOpeningHours(): OpeningHours
+    {
+        $dayOfWeek = (int)date('N');
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/app/src/Model/Store/Store.php:312]: Again, good catch, I have updated it

**Review by grossmannmartin — APPROVED** [review]

**Pull request merged**